tls/s2n_resume.c

```diff
@@ -151,32 +151,90 @@ int s2n_client_serialize_resumption_state(struct s2n_connection *conn, struct s2
-{
-    if (s2n_stuffer_data_available(from) < S2N_STATE_SIZE_IN_BYTES) {
-        POSIX_BAIL(S2N_ERR_INVALID_SERIALIZED_SESSION_STATE);
-    }
-
-    uint8_t format;
-    uint64_t then;
+static S2N_RESULT s2n_tls12_client_deserialize_session_state(struct s2n_connection *conn, struct s2n_stuffer *from)
+{   
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE_REF(from);
 
-    POSIX_GUARD(s2n_stuffer_read_uint8(from, &format));
-    if (format != S2N_TLS12_SERIALIZED_FORMAT_VERSION) {
-        POSIX_BAIL(S2N_ERR_INVALID_SERIALIZED_SESSION_STATE);
-    }
+    RESULT_ENSURE(s2n_stuffer_data_available(from) == (S2N_STATE_SIZE_IN_BYTES - S2N_STATE_FORMAT_LEN), 
+        S2N_ERR_INVALID_SERIALIZED_SESSION_STATE);
 
-    POSIX_GUARD(s2n_stuffer_read_uint8(from, &conn->actual_protocol_version));
+    RESULT_GUARD_POSIX(s2n_stuffer_read_uint8(from, &conn->actual_protocol_version));
 
     uint8_t *cipher_suite_wire = s2n_stuffer_raw_read(from, S2N_TLS_CIPHER_SUITE_LEN);
-    POSIX_ENSURE_REF(cipher_suite_wire);
-    POSIX_GUARD(s2n_set_cipher_as_client(conn, cipher_suite_wire));
+    RESULT_ENSURE_REF(cipher_suite_wire);
+    RESULT_GUARD_POSIX(s2n_set_cipher_as_client(conn, cipher_suite_wire));
 
-    POSIX_GUARD(s2n_stuffer_read_uint64(from, &then));
+    uint64_t then;
+    RESULT_GUARD_POSIX(s2n_stuffer_read_uint64(from, &then));
 
     /* Last but not least, put the master secret in place */
-    POSIX_GUARD(s2n_stuffer_read_bytes(from, conn->secure.master_secret, S2N_TLS_SECRET_LEN));
+    RESULT_GUARD_POSIX(s2n_stuffer_read_bytes(from, conn->secure.master_secret, S2N_TLS_SECRET_LEN));
 
-    return 0;
+    return S2N_RESULT_OK;
+}
+
+static S2N_RESULT s2n_tls13_client_deserialize_session_state(struct s2n_connection *conn, struct s2n_stuffer *from)
+{
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE_REF(from);
+
+    uint8_t protocol_version = 0;
+    RESULT_GUARD_POSIX(s2n_stuffer_read_uint8(from, &protocol_version));
+    ENSURE_EQ(protocol_version, S2N_TLS13);
+
+    uint8_t iana_id[S2N_TLS_CIPHER_SUITE_LEN] = { 0 };
+    RESULT_GUARD_POSIX(s2n_stuffer_read_bytes(from, iana_id, S2N_TLS_CIPHER_SUITE_LEN));
+    struct s2n_cipher_suite *cipher_suite = NULL;
+    GUARD_RESULT(s2n_cipher_suite_from_iana(iana_id, &cipher_suite));
+    ENSURE_REF(cipher_suite);
+
+    uint64_t ticket_issue_time = 0;
+    RESULT_GUARD_POSIX(s2n_stuffer_read_uint64(from, &ticket_issue_time));
+
+    uint32_t ticket_age_add = 0;
+    RESULT_GUARD_POSIX(s2n_stuffer_read_uint32(from, &ticket_age_add));
+
+    uint8_t secret_len = 0;
+    RESULT_GUARD_POSIX(s2n_stuffer_read_uint8(from, &secret_len));
+    ENSURE_LTE(secret_len, S2N_TLS_SECRET_LEN);
+    
+    uint8_t secret[S2N_TLS_SECRET_LEN] = { 0 };
+    RESULT_GUARD_POSIX(s2n_stuffer_read_bytes(from, secret, secret_len));
+
+    /* Construct a PSK from ticket values */
+    struct s2n_psk *psk = NULL;
+    GUARD_RESULT(s2n_array_pushback(&conn->psk_params.psk_list, (void**) &psk));
+    ENSURE_REF(psk);
```

Comment:
Consider s2n_connection_append_psk or s2n_array_insert_and_copy instead. What if this method fails part way through setting up the PSK? You should probably try not to leave the connection in a weird state.

On the other hand, we don't have any existing defense against that in TLS1.2 either. We can set the cipher suite and protocol version but fail before setting the secret. So we could say we just don't support continuing to use a connection if setting the ticket fails.